### PR TITLE
POC: support customised (de)serialiser for edm type.

### PR DESCRIPTION
**Pull request by jjtang1985**
We would like to provide the flexibility so the users can decide for a specific edm type (e.g. Edm.Int32), which js/ts type it will be deserialised to by using which library.
- [x] custom middleware to define a mapping from edm type to (de)seralizer
- [x] when executing the request with your own middleware, the SDK will search for your (de)serialiser first then use default one as fallback.

### CODE REVIEW

File: packages/core/src/odata-v4/payload-value-converter.ts [hunk 1]
```diff
@@ -15,17 +15,25 @@ import { EdmTypeV4 } from './edm-types';
...
-  edmType: EdmTypeShared<'v4'>
+  edmType: EdmTypeShared<'v4'>,
+  customDeserializer?: Partial<EdmTypeMapping>
 ): EdmToPrimitiveV4<T> {
   if (value === null || typeof value === 'undefined') {
     return value;
   }
-  if (deserializers[edmType]) {
-    return deserializers[edmType](value);
+  const customizeDeserializers = getDeserializers(customDeserializer);
```
Comment (jjtang1985) [inline packages/core/src/odata-v4/payload-value-converter.ts:24]: magic to use custom deseralizer.

File: packages/core/test/test-util/test-services/v4/test-service-temporal/TestEntityRequestBuilder.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+  GetAllRequestBuilderV4,
+  EntityV4, Constructable, DestinationOptions, DestinationNameAndJwt, Destination, DTMiddlewareInterface
+} from '../../../../../src';
+import { TestEntity as TestEntityTemporal, TestEntity } from './TestEntity';
+
+/**
+ * Request builder class for operations supported on the [[TestEntity]] entity.
+ */
+export class TestEntityRequestBuilder<T1 = string, T2 = number> extends RequestBuilder<TestEntity<T1, T2>> {
+  // A new class is needed to solve the nested generic type
```
Comment (marikaner) [inline packages/core/test/test-util/test-services/v4/test-service-temporal/TestEntityRequestBuilder.ts:17]: I think you should be able to do this in the requestbuilder directly, when you resolve the generic type there.

**Review by marikaner — COMMENTED** [review]
Just the first few things I stumbled upon.

**Comment by marikaner** [comment]
Example from our discussion:

```ts
// Default entity builder usage
const a = TestEntity.builder().int32Property(1).build();
// Default create request
TestEntity.requestBuilder().create(a).execute();

// Transformed entity builder usage
const b = TestEntity.builder(serializers).int32Property('test').build();
// Transformed create request
TestEntity.requestBuilder(serializers).create(b).execute();
```

File: packages/core/src/datetime-middleware.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
+export interface DTMiddlewareInterface<T1 = string, T2 = number> {
+  deserializers?: {
+    'Edm.String'?: (ori: string) => T1;
+    'Edm.Int32'?: (ori: string) => T2;
+  };
+  serializers?: {
+    'Edm.String'?: (val: T1) => string;
+    'Edm.Int32'?: (val: T2) => string;
+  };
```
[the remark this answers is not in the document]
Comment (jjtang1985) [reply packages/core/src/datetime-middleware.ts]: You are right, the current structure is more sdk developer friendly.

File: packages/core/test/request-builder/v4/get-all-request-builder.spec.ts [hunk 4]
```diff
@@ -72,6 +80,58 @@ describe('GetAllRequestBuilderV4', () => {
...
+      // TestEntityTemporal<string, string>[]
+      const res3 = await TestEntityTemporal.requestBuilder(customMiddlewareFull)
+        .getAll()
+        .execute(defaultDestination);
+      // TestEntityTemporal<string, string>[]
+      const res4 = await TestEntityTemporal.requestBuilder(
+        customMiddlewarePartial
+      )
+        .getAll()
+        .execute(defaultDestination);
```
Comment (jjtang1985) [inline packages/core/test/request-builder/v4/get-all-request-builder.spec.ts:101]: The examples above illustrate how the `requestBuilder()` affects the return type.

File: packages/core/test/request-builder/v4/get-all-request-builder.spec.ts [hunk 5]
```diff
@@ -72,6 +80,58 @@ describe('GetAllRequestBuilderV4', () => {
...
+
+      // TestEntityTemporal<string, number>
+      const entity1 = TestEntityTemporal.builder().int32Property(1).build();
+      // TestEntityTemporal<string, string>
+      const entity2 = TestEntityTemporal.builder(customMiddlewareFull)
+        .int32Property('1')
+        .build();
```
Comment (jjtang1985) [inline packages/core/test/request-builder/v4/get-all-request-builder.spec.ts:108]: With/Without the middleware parameter, see the difference of `int32Property()` that accepts different types.

File: packages/core/test/request-builder/v4/get-all-request-builder.spec.ts [hunk 6]
```diff
@@ -72,6 +80,58 @@ describe('GetAllRequestBuilderV4', () => {
...
+
+      // TestEntityGetAllRequestBuilder<TestEntityTemporal<string, number>, string, number>
+      const requestBuilder1 = await TestEntityTemporal.requestBuilder().getAll();
+      // TestEntityGetAllRequestBuilder<TestEntityTemporal<string, string>, string, string>
+      const requestBuilder2 = await TestEntityTemporal.requestBuilder(
+        customMiddlewareFull
+      ).getAll();
+    });
```
Comment (jjtang1985) [inline packages/core/test/request-builder/v4/get-all-request-builder.spec.ts:116]: Based on the middleware parameter of the `requestBuilder()`, the TestEntityGetAllRequestBuilder has different generic type. 
The `TestEntityCreateRequestBuilder` (not implemented) will rely on the generic type to check the `entity` passed to the `create()` function shown below:
```
TestEntity.requestBuilder(middleware).create(entity).execute();
```

**Review by jjtang1985 — COMMENTED** [review]
I adjusted the code so it reflects the discussion.

**Comment by jjtang1985** [comment]
Hi @artemkovalyov , based on my proposal demonstrated in the meeting, we had new ideas. Please find @marikaner's examples below:

> Example from our discussion:
> 
> ```ts
> // Default entity builder usage
> const a = TestEntity.builder().int32Property(1).build();
> // Default create request
> TestEntity.requestBuilder().create(a).execute();
> 
> // Transformed entity builder usage
> const b = TestEntity.builder(serializers).int32Property('test').build();
> // Transformed create request
> TestEntity.requestBuilder(serializers).create(b).execute();
> ```

You can also find the adjusted code above.

**Comment by jjtang1985** [comment]
I'll close it, because the (ADR)[#934] is merged and this PR should not be merged.

**Pull request closed without merging**